src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+    pub fn output_streams(&mut self) -> Vec<u32> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        self.pcm_infos
+            .as_ref()
+            .unwrap()
+            .iter()
+            .enumerate()
+            .filter(|(_, info)| info.direction == VIRTIO_SND_D_OUTPUT)
+            .map(|(idx, _)| idx as u32)
+            .collect()
+    }
+
+    /// Get all input streams.
+    pub fn input_streams(&mut self) -> Vec<u32> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        self.pcm_infos
+            .as_ref()
+            .unwrap()
+            .iter()
+            .enumerate()
+            .filter(|(_, info)| info.direction == VIRTIO_SND_D_INPUT)
+            .map(|(idx, _)| idx as u32)
+            .collect()
+    }
+
+    /// Get the rates that a stream supports.
+    pub fn rates_supported(&mut self, stream_id: u32) -> Result<PcmRate> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        Ok(self.pcm_infos.as_ref().unwrap()[stream_id as usize]
+            .rate
+            .into())
+    }
+
+    /// Get the formats that a stream supports.
+    pub fn formats_supported(&mut self, stream_id: u32) -> Result<PcmFormats> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        Ok(self.pcm_infos.as_ref().unwrap()[stream_id as usize]
+            .formats
+            .into())
+    }
+
+    /// Get channel range that a stream supports.
+    pub fn channel_range_supported(&mut self, stream_id: u32) -> Result<RangeInclusive<u8>> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        let pcm_info = &self.pcm_infos.as_ref().unwrap()[stream_id as usize];
+        Ok(pcm_info.channels_min..=pcm_info.channels_max)
+    }
+
+    /// Get features that a stream supports.
+    pub fn features_supported(&mut self, stream_id: u32) -> Result<PcmFeatures> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        let pcm_info = &self.pcm_infos.as_ref().unwrap()[stream_id as usize];
+        Ok(pcm_info.features.into())
+    }
+
+    /// Get the latest notification.
+    pub fn latest_notification(&mut self) -> Result<Notification> {
+        if let Some(token) = self.event_queue.peek_used() {
```

Comment:
I removed this function because I couldn't understand the corresponding section in the virtio v1.2 manual about 'the device notifying the driver.' According to the manual, the device should actively send notifications to the event_queue, but my tests showed that it wasn't sending any notifications. Perhaps my implementation of this notification mechanism has some issues.

Reply:
Did you try adding buffers to the event queue for it to write to? It's always up to the driver to allocate buffers, devices can't just add them to the queue out of nowhere.

Reply:
When should I add buffers to the event queue? Should I do it when calling the `latest_notification` function, or during driver initialization? If I add buffers during initialization, how many should I add? I'm not quite clear on this.

Reply:
I'd suggest just adding a single buffer to the event queue during initialisation. `latest_notification` can then add the buffer back to the queue after reading from it.

Reply:
Filling the queue with buffers is also a reasonable option. I've sent #149 with a safe abstraction which may be useful for this.

Reply:
I added some buffers to the `event_queue` for the device to write into and notified the device in `set_up()`, but when calling `latest_notification()`, it seems the device still hasn't written notifications to the previously added buffers. Is there an issue with my implementation?

Reply:
If the implementation to get device notifications is too difficult to grasp, I think it's okay not to implement it. As shown in this [PPT](https://fosdem.org/2024/events/attachments/fosdem-2024-1910-making-virtio-sing-implementing-virtio-sound-in-rust-vmm-project/slides/22175/implementing_virtio-sound_in_rust-vmm_project_HhqbjPY.pdf), they also did not implement an event_queue.

Reply:
I've changed this to use `OwningQueue` to manage the buffers.